### PR TITLE
minimizer consolidation

### PR DESCRIPTION
**Pull request by andped10**
Moved duplicated functionality to base class.

Tests are moved to module for base class.

`minimizer_lmfit` moved public methods to the top of the file.

### CODE REVIEW

File: src/easyscience/fitting/minimizers/minimizer_base.py [hunk 1]
```diff
@@ -166,6 +169,93 @@ def _prepare_parameters(self, parameters: dict[str, float]) -> dict[str, float]:
                     parameters[parameter_name] = item.raw_value
         return parameters
 
+    def _generate_fit_function(self) -> Callable:
```
Comment (andped10) [inline src/easyscience/fitting/minimizers/minimizer_base.py:172]: fetched from bumps / dfo

File: src/easyscience/fitting/minimizers/minimizer_base.py [hunk 2]
```diff
@@ -166,6 +169,93 @@ def _prepare_parameters(self, parameters: dict[str, float]) -> dict[str, float]:
...
+            for constraint in self.fit_constraints():
+                constraint()
+            return_data = func(x)
+            # TODO Loading or manipulating data here
+            return return_data
+
+        return _fit_function
+
+    @staticmethod
+    def _create_signature(parameters: Dict[int, Parameter]) -> Signature:
```
Comment (andped10) [inline src/easyscience/fitting/minimizers/minimizer_base.py:230]: fetched from bumps / lmfit

File: src/easyscience/fitting/minimizers/minimizer_lmfit.py [hunk 3]
```diff
@@ -52,98 +49,21 @@ def __init__(
...
-            # TODO Pre processing here
-            for constraint in self.fit_constraints():
-                constraint()
-            return_data = self._original_fit_function(x)
-            # TODO Loading or manipulating data here
-            return return_data
-
-        lm_fit_function.__signature__ = self._wrap_to_lm_signature(self._cached_pars)
-        return lm_fit_function
+    def available_methods(self) -> List[str]:
```
Comment (andped10) [inline src/easyscience/fitting/minimizers/minimizer_lmfit.py:52]: public method moved to top

File: src/easyscience/fitting/minimizers/minimizer_lmfit.py [hunk 4]
```diff
@@ -52,98 +49,21 @@ def __init__(
         """
         super().__init__(obj=obj, fit_function=fit_function, method=method)
 
-    def _make_model(self, pars: Optional[LMParameters] = None) -> LMModel:
```
Comment (andped10) [inline src/easyscience/fitting/minimizers/minimizer_lmfit.py:55]: private method moved down

File: tests/unit_tests/Fitting/minimizers/test_minimizer_bumps.py [hunk 5]
```diff
@@ -91,16 +90,9 @@ def test_generate_fit_function(self, minimizer: Bumps) -> None:
 
         # Then
         fit_function = minimizer._generate_fit_function()
-        fit_function_result = fit_function([10.0])
```
Comment (andped10) [inline tests/unit_tests/Fitting/minimizers/test_minimizer_bumps.py:94]: now tested in `minimizer_base`

File: tests/unit_tests/Fitting/minimizers/test_minimizer_lmfit.py [hunk 6]
```diff
@@ -361,22 +331,3 @@ def test_gen_fit_results(self, minimizer: LMFit, monkeypatch) -> None:
         assert str(domain_fit_results.minimizer_engine) == "<class 'easyscience.fitting.minimizers.minimizer_lmfit.LMFit'>"
         assert domain_fit_results.fit_args is None
 
-    def test_wrap_to_lm_signature(self, minimizer: LMFit) -> None:
```
Comment (andped10) [inline tests/unit_tests/Fitting/minimizers/test_minimizer_lmfit.py:364]: now tested in `minimizer_base`

File: src/easyscience/fitting/minimizers/minimizer_lmfit.py [hunk 7]
```diff
@@ -313,50 +283,3 @@ def _gen_fit_results(self, fit_results: ModelResult, **kwargs) -> FitResults:
...
-            'lbfgsb',
-            'powell',
-            'cg',
-            'newton',
-            'cobyla',
-            'bfgs',
-        ]
-
-    @staticmethod
-    def _wrap_to_lm_signature(parameters: Dict[int, Parameter]) -> Signature:
```
Comment (andped10) [inline src/easyscience/fitting/minimizers/minimizer_lmfit.py:334]: now in `minimizer_base`

File: tests/unit_tests/Fitting/minimizers/test_minimizer_lmfit.py [hunk 8]
```diff
@@ -87,56 +84,29 @@ def test_make_model_no_pars(self, minimizer: LMFit, monkeypatch) -> None:
         assert mock_lm_model.set_param_hint.call_count == 2
         assert model == mock_lm_model
 
-    def test_generate_fit_function_signatur(self, minimizer: LMFit) -> None:
```
Comment (andped10) [inline tests/unit_tests/Fitting/minimizers/test_minimizer_lmfit.py:90]: now tested in `minimizer_base`

File: src/easyscience/fitting/minimizers/minimizer_lmfit.py [hunk 9]
```diff
@@ -257,6 +177,56 @@ def convert_to_par_object(parameter: Parameter) -> LMParameter:
...
+    def _generate_fit_function(self) -> Callable:
+        """
+        Using the user supplied `fit_function`, wrap it in such a way we can update `Parameter` on
+        iterations.
+
+        :return: a fit function which is compatible with lmfit models
+        """
+        fit_function = super()._generate_fit_function()
+        fit_function.__signature__ = self._create_signature(self._cached_pars)
+        return fit_function
```
Comment (rozyczko) [inline src/easyscience/fitting/minimizers/minimizer_lmfit.py]: `_generate_fit_function` here and in bumps have the same code. They call base class implementation then add signature based on `_cached_pars` which is also accessible in the base class. Should these be just merged with the base class method?

**Review by rozyczko — APPROVED** [review]
Clearly done refactoring, one minor question raised.
Ready to merge.

File: src/easyscience/fitting/minimizers/minimizer_lmfit.py (hunk 9 above)
Comment (andped10) [reply src/easyscience/fitting/minimizers/minimizer_lmfit.py]: You are right.  Moved the signature generation to the base class and adjusted tests

**Pull request merged**